src/map/status.c

```diff
@@ -11114,18 +11263,192 @@ static int status_change_end_(struct block_list *bl, enum sc_type type, int tid,
 					break;
 			}
 			break;
-		case SC_ARMOR_RESIST:
-			if (sce->val1 > 0)
-				clif->sc_end(bl, bl->id, AREA, status->get_sc_icon(SC_RESIST_PROPERTY_WATER));
+		case SC_ARMOR_RESIST: {
+			struct status_change_entry *sce_water = sc->data[SC_RESIST_PROPERTY_WATER];
+			struct status_change_entry *sce_ground = sc->data[SC_RESIST_PROPERTY_GROUND];
+			struct status_change_entry *sce_fire = sc->data[SC_RESIST_PROPERTY_FIRE];
+			struct status_change_entry *sce_wind = sc->data[SC_RESIST_PROPERTY_WIND];
 
-			if (sce->val2 > 0)
-				clif->sc_end(bl, bl->id, AREA, status->get_sc_icon(SC_RESIST_PROPERTY_GROUND));
+			// Water
+			clif->sc_end(bl, bl->id, AREA, status->get_sc_icon(SC_RESIST_PROPERTY_WATER));
+			if (sce_water != NULL) {
+				int sum_water = sce_water->val1 + ((sce_fire != NULL) ? sce_fire->val2 : 0);
+				if (sce_water->val1 > 0 && sum_water > 0) {
+					const struct TimerData *td = timer->get(sce_water->timer);
+					int sc_icn = status->get_sc_icon(SC_RESIST_PROPERTY_WATER);
+					int sc_typ = status->get_sc_relevant_bl_types(SC_RESIST_PROPERTY_WATER);
+					int sc_tck = (int)DIFF_TICK(td->tick, timer->gettick());
+					int sc_ttl = sce_water->total_tick;
+					clif->status_change_sub(bl, sc_icn, sc_typ, 1, sc_tck, sc_ttl, 0, 0, 0);
+				}
+			}
 
-			if (sce->val3 > 0)
-				clif->sc_end(bl, bl->id, AREA, status->get_sc_icon(SC_RESIST_PROPERTY_FIRE));
+			// Ground
+			clif->sc_end(bl, bl->id, AREA, status->get_sc_icon(SC_RESIST_PROPERTY_GROUND));
+			if (sce_ground != NULL) {
+				int sum_ground = sce_ground->val1 + ((sce_wind != NULL) ? sce_wind->val2 : 0);
+				if (sce_ground->val1 > 0 && sum_ground > 0) {
+					const struct TimerData *td = timer->get(sce_ground->timer);
+					int sc_icn = status->get_sc_icon(SC_RESIST_PROPERTY_GROUND);
+					int sc_typ = status->get_sc_relevant_bl_types(SC_RESIST_PROPERTY_GROUND);
+					int sc_tck = (int)DIFF_TICK(td->tick, timer->gettick());
+					int sc_ttl = sce_ground->total_tick;
+					clif->status_change_sub(bl, sc_icn, sc_typ, 1, sc_tck, sc_ttl, 0, 0, 0);
+				}
+			}
 
-			if (sce->val4 > 0)
-				clif->sc_end(bl, bl->id, AREA, status->get_sc_icon(SC_RESIST_PROPERTY_WIND));
+			// Fire
+			clif->sc_end(bl, bl->id, AREA, status->get_sc_icon(SC_RESIST_PROPERTY_FIRE));
+			if (sce_fire != NULL) {
+				int sum_fire = sce_fire->val1 + ((sce_ground != NULL) ? sce_ground->val2 : 0);
+				if (sce_fire->val1 > 0 && sum_fire > 0) {
+					const struct TimerData *td = timer->get(sce_fire->timer);
```

Comment:
``timer->get`` may return NULL. also sce_fire->timer can be invalid timer. probably need add checks for timer field and return value of ``timer->get``?